File: beer-recommender/python/recommender.py
import pandas as pd
import numpy as np
from sklearn.metrics.pairwise import cosine_similarity
import os


filename = os.path.join("data/beer_reviews.csv")
df = pd.read_csv(filename)
# let's limit things to the top 250
n = 250
top_n = df.beer_name.value_counts().index[:n]
df = df[df.beer_name.isin(top_n)]

print df.head()
print "melting..."
df_wide = pd.pivot_table(df, values=["review_overall"],
                         rows=["beer_name", "review_profilename"],
                         aggfunc=np.mean).unstack()

# any cells that are missing data (i.e. a user didn't buy a particular product)
# we're going to set to 0
df_wide = df_wide.fillna(0)

# this is the key. we're going to use cosine_similarity from scikit-learn
# to compute the distance between all beers
print "calculating similarity"
dists = cosine_similarity(df_wide)

# stuff the distance matrix into a dataframe so it's easier to operate on
dists = pd.DataFrame(dists, columns=df_wide.index)

# give the indicies (equivalent to rownames in R) the name of the product id
dists.index = dists.columns


def get_sims(products):
    """
    get_top10 takes a distance matrix an a productid (assumed to be integer)
    and will calculate the 10 most similar products to product based on the
    distance matrix

    dists - a distance matrix
    product - a product id (integer)
    """
    p = dists[products].apply(lambda row: np.sum(row), axis=1)
    p = p.order(ascending=False)
    return p.index[p.index.isin(products) == False]


get_sims(["Sierra Nevada Pale Ale", "120 Minute IPA", "Stone Ruination IPA"])

from yhat import Yhat, YhatModel, preprocess


class BeerRecommender(YhatModel):
    @preprocess(in_type=dict, out_type=dict)
    def execute(self, data):
        beers = data.get("beers")
        suggested_beers = get_sims(beers)
        result = []
        for beer in suggested_beers:
            result.append({"beer": beer})
        return result


yh = Yhat("YOUR_USERNAME", "YOUR_APIKEY", "http://cloud.yhathq.com/")

if raw_input("Deploy? (y/N)") == "y":
    print yh.deploy("BeerRecommender", BeerRecommender, globals())

print yh.predict("BeerRecommender", {"beers": ["Sierra Nevada Pale Ale",
                 "120 Minute IPA", "Stone Ruination IPA"]})

File: beer-recommender/r/recommender.r
library(reshape2)
library(plyr)

df <- read.csv("./data/beer_reviews.csv")

head(df)

n <- 100
beer_counts <- table(df$beer_name)
beers <- names(beer_counts[beer_counts > n])
df <- df[df$beer_name %in% beers,]

df.wide <- dcast(df, beer_name ~ review_profilename,
            value.var='review_overall', mean, fill=0)
head(df.wide)
dists <- dist(df.wide[,-1], method="euclidean")
dists <- as.data.frame(as.matrix(dists))
colnames(dists) <- df.wide$beer_name
dists$beer_name <- df.wide$beer_name
head(dists)


getSimilarBeers <- function(beers_i_like) {
  beers_i_like <- as.character(beers_i_like)
  cols <- c("beer_name", beers_i_like)
  best.beers <- dists[,cols]
  if (ncol(best.beers) > 2) {
    best.beers <- data.frame(beer_name=best.beers$beer_name, V1=rowSums(best.beers[,-1]))
  }
  results <- best.beers[order(best.beers[,-1]),]
  names(results) <- c("beer_name", "similarity")
  results[! results$beer_name %in% best.beers,]
}
getSimilarBeers(c("Coors Light"))

model.transform <- function(df) {
  df
}

model.predict <- function(df) {
  getSimilarBeers(df$beers)
}

testcase <- data.frame(beers=c("Coors Light"))
model.predict(testcase)


library(yhatr)
yhat.config <- c(username="{ USERNAME }", apikey="{ APIKEY }", env="http://sandbox.yhathq.com/")
yhat.deploy("BeerRecR")

# curl -X POST -H "Content-type: application/json" \
#   --user { USERNAME }:{ APIKEY } \
#   --data '{"beers": ["Coors Light"]}' \
#   http://sandbox.yhathq.com/{ USERNAME }/models/BeerRecR/
# {
#   "yhat_id": "6f232a20-5830-4410-9ab3-362694d52f38",
#   "result": {
#     "beer_name": [
#       "Michelob Ultra",
#       "Natural Light",
#       "Blue Moon Belgian White",
#       "Fat Tire Amber Ale",
#       "Dale's Pale Ale",
#       "Guinness Draught",
#       "60 Minute IPA",
#       "Sierra Nevada Pale Ale"
#       ],
#     "similarity": [
#       93.80931,
#       101.99939,
#       142.29593,
#       161.32967,
#       163.03715,
#       165.23529,
#       205.22244,
#       207.33894
#       ]
#   },
#   "yhat_model": "BeerRecR"
# }


